### ISSUE TITLE
[QUESTION] How do I use a string-formatted BaseModel as a path param?

### ISSUE THREAD

**Issue by AllSeeingEyeTolledEweSew** [tiangolo/fastapi#3059]
### Example

```Python
from fastapi import FastAPI
from pydantic import BaseModel

class Multihash(BaseModel):
    func: int
    digest: bytes

class MultihashHex(Multihash):
    @classmethod
    def __modify_schema__(cls, field_schema: Dict[str, Any]) -> None:
        field_schema.update({"type": "string", "format": "multihash-hex"})

    @classmethod
    def __get_validators__(cls) -> Iterator[Callable[..., Any]]:
        yield cls.validate

    @classmethod
    def validate(cls, value: Union[str, bytes]) -> Multihash:
        if isinstance(value, str):
            value = bytes.fromhex(value)
        return Multihash(...)


app = FastAPI()

@app.get("/data/{mh}")
def read(mh:MultihashHex):
    return ...
```

### Description

When I run the above, I get:
```
$ uvicorn test:app
...
  File "./test.py", line 32, in <module>
    def read(mh:MultihashHex):
  File "/home/amnesia/Persistent/venv/lib/python3.7/site-packages/fastapi/routing.py", line 574, in decorator
    callbacks=callbacks,
  File "/home/amnesia/Persistent/venv/lib/python3.7/site-packages/fastapi/routing.py", line 520, in add_api_route
    callbacks=current_callbacks,
  File "/home/amnesia/Persistent/venv/lib/python3.7/site-packages/fastapi/routing.py", line 380, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
  File "/home/amnesia/Persistent/venv/lib/python3.7/site-packages/fastapi/dependencies/utils.py", line 308, in get_dependant
    ), "Path params must be of one of the supported types"
AssertionError: Path params must be of one of the supported types
```

I'm building an app that vends content-addressable data. I'd like to use a [multihash](https://multiformats.io/multihash/) as a path param. Multihash is just a (hash-function, hash-digest) tuple, with well-defined encoding.

My intent is to have my path param be a hexadecimal multihash, and use the `Multihash` type internally. There seem to be other examples of string-representable objects, like `NameEmail` or `IPv4Address`.

But, I can't seem to make my `Multihash` class be a `BaseModel`. I get the above error.

I can make `Multihash` a base `object` instead, but I lose pydantic features.

Is it an error that `BaseModel` objects can't be used this way, or is there a better way to do what I'm doing?

### Environment

* OS: tails 4.17
* python: 3.7.3
* packages:
```
fastapi==0.63.0
pydantic==1.8.1
```

**Comment by insomnes** [comment]
@AllSeeingEyeTolledEweSew hello!
Maybe you can just set it as str and model it in your path function:
```python3
from fastapi import FastAPI, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, ValidationError

app = FastAPI()


class TestModel(BaseModel):
    value: int


@app.get("/{test_str}")
def main(test_str: str):
    try:
        modeled = TestModel(value=test_str)
    except ValidationError as exc:
        return JSONResponse(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            content={"detail": exc.errors()}
        )

    return modeled

```

**Comment by AllSeeingEyeTolledEweSew** [comment]
> Maybe you can just set it as str and model it in your path function:

Right, that's another option I forgot to mention. The downside is it's repetitive as I add more path functions.

And it seems the _spirit_ of FastAPI and pydantic is to need _less_ parsing and validation in application code, so it seems there should be a better answer than this.

**Comment by insomnes** [comment]
@AllSeeingEyeTolledEweSew 
>Right, that's another option I forgot to mention. The downside is it's repetitive as I add more path functions.

You can use dependency injection system for that.

I can only clearly say, that it's not an error, as you can see here in the docs:
https://fastapi.tiangolo.com/tutorial/body/#request-body-path-query-parameters

And in code of `is_scalar_field` function in `utils.py`

**Comment by AllSeeingEyeTolledEweSew** [comment]
> as you can see here in the docs:
> https://fastapi.tiangolo.com/tutorial/body/#request-body-path-query-parameters

Reading that:
> If the parameter is also declared in the path, it will be used as a path parameter.
> If the parameter is of a singular type (like int, float, str, bool, etc) it will be interpreted as a query parameter.
> If the parameter is declared to be of the type of a Pydantic model, it will be interpreted as a request body.

Reading it in order, I thought it says plainly that if a parameter is declared in the path then it's used as a path parameter, _even if it's a `BaseModel`_.

If it _intends_ to say that "`BaseModel`s will _always_ be interpreted as request body", then... I feel it should be written differently :)

I know `is_scalar_field` is part of the stack that blocks this usage. But surely an undocumented utility function is not ground truth for APIs or intent

Extending my example from above:

```python
class Outer(BaseModel):
    mh: MultihashHex

parse_raw_as(Outer, '{"mh": "0a2b3c..."}')
"""
Outer(mh=MultihashHex(...))
"""
```

Pydantic lets me parse a string field into a `BaseModel` just fine, so I expect to be able to parse a path parameter into a `BaseModel` as well. The two seem perfectly analogous to me

**Comment by hiboo** [comment]
Came to the exact same issue when I wanted to provide a path parameter which is an object with several keys. As mentioned previously, a test in source code prevent us from being able to provide something else than scalars as path parameters... 

However, the OpenAPI specification currently allow and define a way to document path parameters as objects or arrays thanks to `schema`, `style` and `explode` properties of path parameter objects, see https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#parameter-object

IMO, It could be interesting to implement such a thing
